lib/redis_memo/memoize_query/invalidation.rb

```diff
@@ -123,50 +123,89 @@ def self.rewrite_import_method(model_class, method_name)
         # - default values filled by the database
         # - updates on conflict conditions
         records_to_invalidate =
-          if unique_by
+          if columns_to_update
             RedisMemo::MemoizeQuery::Invalidation.send(
-              :select_by_uniq_index,
+              :select_by_columns,
               records,
-              unique_by,
+              columns_to_update,
             )
           else
             []
           end
 
         result = send(:"#{method_name}_without_redis_memo_invalidation", *args, &blk)
 
-        records_to_invalidate += RedisMemo.without_memo do
-          # Not all databases support "RETURNING", which is useful when
-          # invaldating records after bulk creation
-          model_class.where(model_class.primary_key => result.ids).to_a
+        # Offload the records to invalidate while selecting the next set of
+        # records to invalidate
+        case records_to_invalidate
+        when Array
+          RedisMemo::MemoizeQuery.invalidate(*records_to_invalidate) unless records_to_invalidate.empty?
+
+          RedisMemo::MemoizeQuery.invalidate(*RedisMemo::MemoizeQuery::Invalidation.send(
+            :select_by_id,
+            model_class,
+            # Not all databases support "RETURNING", which is useful when
+            # invaldating records after bulk creation
+            result.ids,
+          ))
+        else
+          RedisMemo::MemoizeQuery.invalidate_all(model_class)
         end
 
-        memos_to_invalidate = records_to_invalidate.map do |record|
-          RedisMemo::MemoizeQuery.to_memos(record)
-        end
-        RedisMemo::Memoizable.invalidate(memos_to_invalidate.flatten)
-
         result
       end
     end
   end
 
-  def self.select_by_uniq_index(records, unique_by)
+  def self.select_by_columns(records, columns_to_update)
     model_class = records.first.class
     or_chain = nil
-
-    records.each do |record|
-      conditions = {}
-      unique_by.each do |column|
-        conditions[column] = record.send(column)
+    columns_to_select = columns_to_update & RedisMemo::MemoizeQuery
+      .memoized_columns(model_class)
+      .to_a.flatten.uniq
+
+    # Nothing to invalidate here
+    return [] if columns_to_select.empty?
+
+    RedisMemo::Tracer.trace(
+      'redis_memo.memoize_query.invalidation',
+      "#{__method__}##{model_class.name}",
+    ) do
+      records.each do |record|
+        conditions = {}
+        columns_to_select.each do |column|
+          conditions[column] = record.send(column)
+        end
+        if or_chain
+          or_chain = or_chain.or(model_class.where(conditions))
+        else
+          or_chain = model_class.where(conditions)
+        end
       end
-      if or_chain
-        or_chain = or_chain.or(model_class.where(conditions))
+
+      record_count = RedisMemo.without_memo { or_chain.count }
+      if record_count > bulk_operations_invalidation_limit
```

Comment:
I wonder if it'd be better to just look at the total # of records being imported, to avoid this extra query / extra computation to iterate through all the records.

Reply:
> extra computation to iterate through all the records.

Well this will send a `SELECT count(*)` query.

By sending this query we can avoid the expensive payload transfer and activerecord record installation.

> I wonder if it'd be better to just look at the total # of records being imported

I don't think this would do. For example, if we `on_duplicate_key_update: [:visibility]`, and we are also memorizing `visibility`, we could still be getting a lot of records back from the database

Reply:
Actually, now that I think about it, we can still query by uniq_by if we're really querying by uniq_by instead of those columns to update! That should be more efficient and we don't need to worry about the size limit since the uniq_by set is smaller or equal to the record set we're importing.

Reply:
nvm -- we cannot really query by `uniq_by` only because the uniq_by value might or might not exist -- it could be something filled by the database.  So yeah I think query by the columns_to_update is still the best we can do.

Reply:
> the best we can do

not really the best we can do, but I think it's good enough for now. Let's save more optimizations for later